### ISSUE TITLE
[FR] Support garch style models on sigma

### ISSUE THREAD

**Issue by SteveBronder** [paul-buerkner/brms#843]
## Summary

It would be cool to be able to model `sigma` as a garch process.

## Description

After closing #708, brms can support autocor processes in the formula syntax ala

```stan
fit <- brm(y ~ arma(p = 2, q = 2), data = list(y = y),
           inits = 0)
```

Though this is only available on the main terms (Paul is that right?) and not distribution parameters.

For some parameters there are specific autocorrelation models like the [General Autoregressive Conditional Heteroskedasticity](https://vlab.stern.nyu.edu/docs/volatility) (garch) and general volatility models. Garch are used pretty frequently there are several stan implementations of them such as [varstan](https://github.com/asael697/varstan/blob/master/inst/stan/garch.stan). So I think it would be nice to add these given the new formula syntax

## Example 

For a model written with either `student` or `normal` distributions like

```R
fit <- brm(y ~ arma(p = 2, q = 2),
      sigma ~ garch(p = 2, rr =2), data = list(y = y), inits = 0)
```

The stan code could look something like

```stan
  for (n in 1:N) {
   // standard arma stuff
    mu[n] += Err[n, 1:Kma] * ma;
    err[n] = Y[n] - mu[n];
    for (i in 1:J_lag[n]) {
      Err[n + 1, i] = err[n + 1 - i];
    }
    mu[n] += Err[n, 1:Kar] * ar;
    // garch stuff
    // pretend we have a pow that does elementwise
    sigma[n] += pow(Err[n, 1:GKar], 2) * g_ar;  
    for (i in 1:G_lag[n]) {
      sigma_lag[n + 1, i] = sigma[n + 1 - i];
    }
    sigma[n] += pow(sigma_lag[n, 1:GKrr], 2) * g_rr;
  }
```

Some Qs

1. Is it okay if these are just limited to normal and student (what does arma currently support for cov=FALSE)?
2. Do you want a list of other parameterized models that would probably be reasonable to build out? Would things like bsts be in scope of that?

If you have the time to write an outline on what needs to be done to add new autocor I wouldn't mind taking a shot at adding garch and some of it's variants. I posted about this before but imo I think this would also be a good GSOC or summer intern problem. Either way, if we can get the garch one out in the next few months I can probably do a talk about it at R in Finance to get some other contributors

**Comment by paul-buerkner** [comment]
Thanks for opening this issue! 

Correlation structures are already technically implemented for all parameters, but most are prohibited for non-main parameters as they don't make much sense there. So including garch terms on `sigma` shouldn't be too difficult.

> Is it okay if these are just limited to normal and student (what does arma currently support for cov=FALSE)?

Yes. These are the currently the only supported families supporting ARMA for `cov = FALSE`.

> Do you want a list of other parameterized models that would probably be reasonable to build out? 

That would be nice! Perhaps in a new issue.

> Would things like bsts be in scope of that?

I am not entirely sure. I had an implementation for simple BSTS models in brms but (the implementation of) it didn't quite fit into brms. So I deprecated and later removed it. If we can find a way to express these models without causing too much trouble in maintaining them and making them work with most other brms features, I think it could be worth implementing them again.

> If you have the time to write an outline on what needs to be done to add new autocor I wouldn't mind taking a shot at adding garch and some of it's variants. I posted about this before but imo I think this would also be a good GSOC or summer intern problem.

I will take GARCH models as an opportunity to write down what needs to be done, as I am doing it, since this is the first time I am implementing a new correlation structure in the new formula based framework.

**Comment by SteveBronder** [comment]
Paul have you had any time to write this stuff down? I know you are v busy so no worries!

**Comment by paul-buerkner** [comment]
I have not unfortunately. Things are moving a little slower these days. I see when I can manage to implement it.

**Comment by paul-buerkner** [comment]
Two quick questions:

- From a quick look at the Stan manual, it seems GARCH models are defined on variances. Does the above code take into account that `sigma` is the SD not the variance?
- Is there a specific reason, to call the argument `rr`, rather than, say, `q` as in the ARMA case?

**Comment by crossvalidated** [comment]
This would be a  very useful feature. Maybe this video about a Stan implementation could help with the implementation? https://www.youtube.com/watch?v=nwuU-KEKXhU&t=23s

**Comment by SteveBronder** [comment]
> From a quick look at the Stan manual, it seems GARCH models are defined on variances. Does the above code take into account that sigma is the SD not the variance?

Yes that's why it has the `pow(x, 2)` in it. Though I should probably double check this @asael697 does that sound right?

> Is there a specific reason, to call the argument rr, rather than, say, q as in the ARMA case?

It's just what they call it in the lit

> How do we initialize sigma? I guess there needs to be some initial parameter for this purpose, or not?

Like `sigma[1]`? I'd check out @asael697's garch impl [here](https://github.com/asael697/bayesforecast/blob/master/inst/stan/tgarch.stan#L103). There he uses a `sigma0` and `mu0` with a prior on the for the initial parameter

> How does a GARCH(p, q) model look like if we don't have an ARMA model defined on mu at the same time?

I have to double check this, but looking at the [wiki](https://en.wikipedia.org/wiki/Autoregressive_conditional_heteroskedasticity#GARCH(p,_q)_model_specification) wouldn't that just be the standard case of 

![image](https://user-images.githubusercontent.com/5857231/103034202-5fa0ff00-4532-11eb-9ff2-c4d8eb31b552.png)

i.e. the mean is 0?

**Comment by asael697** [comment]
I gotta say I love the idea, the fable package does multilevel time series analysis, with a similar syntaxis will be nice that It will be established in brms or another package.

> Yes that's why it has the `pow(x, 2)` in it. Though I should probably double check this @asael697 does that sound right?
 The code looks ok..

Just this I am not sure:
    mu[n] += Err[n, 1:Kar] * ar;
If you apply to the mean the autoregressive coefficients the you have to update ERR again like ERR = y -mu . Cause thee garch is using the process' errors. Also I will use  s and k instead of p and q in the **brms syntax**

> > Is there a specific reason, to call the argument rr, rather than, say, q as in the ARMA case?
> 
> It's just what they call it in the lit
I also think is the process errors. If it has a location-scale model then can be re-writeen as
$y_t = \mu_t +\sigma_t* Err_t$

> Like `sigma[1]`? I'd check out @asael697's garch impl [here](https://github.com/asael697/bayesforecast/blob/master/inst/stan/tgarch.stan#L103). There he uses a `sigma0` and `mu0` with a prior on the for the initial parameter
 **Yes**
> > How does a GARCH(p, q) model look like if we don't have an ARMA model defined on mu at the same time?
> 
> I have to double check this, but looking at the [wiki](https://en.wikipedia.org/wiki/Autoregressive_conditional_heteroskedasticity#GARCH(p,_q)_model_specification) wouldn't that just be the standard case of

you can do it in bayesforecast with garch(ts,order = (s,k,0)) in this case will look like these 2 equations:
 
$$y_t = \mu_0 + \sigma_t \epsilon_t \text{ for all } t \in T$$
$$\sigma_t = \sigma_0 + \sum^s \alpha_i \epsilon_{t-i}^2 + \sum^k \beta_i \sigma_{t-i}^2$$

NO equations for the mean, For the code above,  more IF statements will do the magic :)

**Comment by paul-buerkner** [comment]
I think with the existing internal brms interfaces in place, it would be a good project to pursue.

**Comment by SteveBronder** [comment]
Awesome! @paul-buerkner and @asael697 would you both have a bit of time to help out the GSoC student if we get one? If so I can setup the base of the docs we need and we can chat over the phone some time to sort out the steps of the project. I think it might be something like

1. Write Stan examples for a few of the garch stye models [here](https://vlab.stern.nyu.edu/docs/volatility) and use SBC to show they can be well calibrated in a Bayesian setting. I think writing a few of these will help break out the different mechanical parts of each model into a more general form.
2. Write tech spec as an issue in BRMS suggesting the syntax style, supported models, outstanding issues, any good default priors for the parameters that were found, etc. (Paul you'd probably need to lay out the template here).
3. Prototype and tests for simple `garch(p, q)` on sigma
4. Add additional garch flavors

I like each of these steps because successfully completing any of them should at least drive this forward

**Comment by asael697** [comment]
Hola,

Yes count with me!  some variants of GARCH are already implemented  in bayesforecast I can help with that part! :)

Greetings,
Asael

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by paul-buerkner** [comment]
Sure! I am in!

Asael A Matamoros <notifications@github.com> schrieb am Fr., 22. Jan. 2021,
22:25:

**Comment by DoktorMike** [comment]
Just bumping on this since I'm curious? Were you able to find a GSoC student?
